providers/keycloak_oidc.go

```diff
@@ -0,0 +1,132 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
+)
+
+const keycloakOIDCProviderName = "Keycloak OIDC"
+
+// KeycloakOIDCProvider creates a Keycloak provider based on OIDCProvider
+type KeycloakOIDCProvider struct {
+	*OIDCProvider
+}
+
+// NewKeycloakOIDCProvider makes a KeycloakOIDCProvider using the ProviderData
+func NewKeycloakOIDCProvider(p *ProviderData) *KeycloakOIDCProvider {
+	p.ProviderName = keycloakOIDCProviderName
+	return &KeycloakOIDCProvider{
+		OIDCProvider: &OIDCProvider{
+			ProviderData: p,
+		},
+	}
+}
+
+var _ Provider = (*KeycloakOIDCProvider)(nil)
+
+// AddAllowedRoles sets Keycloak roles that are authorized.
+// Assumes `SetAllowedGroups` is already called on groups and appends to that
+// with `role:` prefixed roles.
+func (p *KeycloakOIDCProvider) AddAllowedRoles(roles []string) {
+	if p.AllowedGroups == nil {
+		p.AllowedGroups = make(map[string]struct{})
+	}
+	for _, role := range roles {
+		p.AllowedGroups[formatRole(role)] = struct{}{}
+	}
+}
+
+// EnrichSession is called after Redeem to allow providers to enrich session fields
+// such as User, Email, Groups with provider specific API calls.
+func (p *KeycloakOIDCProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	err := p.OIDCProvider.EnrichSession(ctx, s)
+	if err != nil {
+		return err
+	}
+	return p.extractRoles(ctx, s)
+}
+
+func (p *KeycloakOIDCProvider) extractRoles(ctx context.Context, s *sessions.SessionState) error {
+	claims, err := p.getAccessClaims(ctx, s)
+	if err != nil {
+		return err
+	}
+
+	var roles []string
+	roles = append(roles, claims.RealmAccess.Roles...)
+	roles = append(roles, getClientRoles(claims)...)
+
+	// Add to groups list with `role:` prefix to distinguish from groups
```

Comment:
@JoelSpeed Something I changed here that deviates from a convention we started in the `gitlab` provider that I think I want to change as well:

When there are multiple group/role types to populate our session.Groups list -- only things that aren't explicitly already a `group` will get a prefix (e.g. `role:blah`). Leaving the groups raw instead of `group:somegroup`.

This way, we can fully rely the base `OIDCProvider` logic for standard groups extraction without having to make each Provider we use this new pattern of extending it have to override that base logic.

Reply:
If you concur, before we have the next major release, I'd like to sneak into the `gitlab` provider and make this shift.  I think the `gitlab` provider could use a refactor to be based on the `OIDCProvider` like this PR to remove alot of duplicate code it has.

Reply:
Are there any providers that we know of that allow `:` as a character in the group name, if there are, then a user could theoretically create a group called `role:admin`, which would then clash with a role called `admin`. If we explicitly prefix within  the groups as well, I don't think this confusion is possible.

I'm concerned this will one day turn into a weird bug where a user ends up in a role because of an ill named group :thinking:

Reply:
Hi, I understand your concerns and it might pose a risk of unintended behavior.
Still I think that especially for Keycloak there is a difference between group membership and roles (realm and client level).

In our usecase we use roles for permissions -> for this proxy this means „you can pass or you shall not pass“ and we use groups as a manner of managing users (all members of a group have the same permissions but multiple groups can have those permissions).

I have seen other users and implementations that use the „groups“ for permissions.

On top of that at least Keycloak will allow us to design the tokens the way we want to… so there is a lot that can be done wrong.

The implementation of @pb82 leverages the default behavior of Keycloak and where it would map groups and roles into the token. So I think for Keycloak it is alright to use „role:“ as a delimiter. But: Keycloak does allow me to create a group „role:admin“ and I am not aware of any way around this.
One Idea would be to configure if we want to authorize based upon groups or on roles.

One extr thing, that I could imagine is that, if we went for a completely independent OIDC JWT implementation we could configure a JSON Path to the claim in the token, we want to verify and then configure what values would be required to let the user pass.

Reply:
Good call out @JoelSpeed!

While this isn't externally forceable, if a misconfiguration did occur this would be an authz bypass vulnerability for some installations.

I want to figure out a way where the simple behavior in most provider usages that just use `group` don't have a `group:blah` prefix in every group in the `X-Forwarded-Groups` header. Not forcing the `group` canonical use case if a small percentage of the sysadmins use a corner case authz mechanism (e.g. Keycloak roles, or GitLab projects).

Maybe we figure out a scheme that is clear, but uses characters that are very likely not allowed in group names (or wouldn't be common)?

Maybe borrowing type casting syntax? `(role)admin` instead of `role:admin`?

Reply:
I had wondered if we set some sort of option on the OIDC provider to prefix or not to prefix. Then where we have specific providers that have these multiple types, groups, roles, projects, that provider would say it needs groups prefixed and that would affect all installations for that provider. 

To me that's a reasonable limitation when using a provider that supports multiple types of grouping. The usage would just be that when using allowed groups you have to set `<type>:<name>` always on certain providers (we could have validation for this)

I haven't looked at the code though, @NickMeves do you think this would be achievable?

Reply:
> I had wondered if we set some sort of option on the OIDC provider to prefix or not to prefix.

That's a good idea, would potentially help with each sub-provider that has customizations needing to override `Authorization` based logic (I think GitLab needed to do this). Vs sticking with pure no-prefix OIDC group logic, I could keep the code simple and not override default implementations.

Doesn't help the header bloat with people who only use `groups` now having `X-Forwarded-Groups: group:foo,group:bar` -- but maybe that's just cosmetic and isn't too annoying...?

My fear: It might impact OOTB vendor software that takes in comma-delimited group names from a header to internal RBAC mappings/assignments (the original intent of `X-Forwarded-Groups`) with `group:` prefixed 🤔

Reply:
Or crazy idea we ditch the monolithic `session.Groups` and start splitting off various grouping concepts in each provider into their own named header: `X-Forwarded-Groups`, `X-Forwarded-Roles`, `X-Forwarded-Projects`, `X-Forwarded-Repos`, etc...

But that's a much bigger change 😰

Reply:
> My fear: It might impact OOTB vendor software that takes in comma-delimited group names from a header to internal RBAC mappings/assignments (the original intent of `X-Forwarded-Groups`) with `group:` prefixed 🤔

My wonky type casting idea would definitely break this. We should stick with `prefix:name` style to be compatible with downstream consuming syntax.

Reply:
Other option is warning log messages -- if we detect an un-prefixed group that matches a registered alternative for a given provider, we emit a warning log message?

Reply:
@JoelSpeed -- I think I've talked myself into leaving the formatting as is and not punishing 99% of users for 1% who might make weird configuration choices that cause conflicts (which is in their power as admins to remediate). Especially since this is purely an admin setup issue, and can't be forced maliciously from client-side (hence no large security issue).

I think the user experience for people who use backend systems who will take groups from headers and convert to backend systems' RBAC is much better if those backend systems don't have `group:` in every name assuming default `group` structure from OIDC IDTokens is the primary use case.

Especially if the backend system doesn't support `:` in their group/role structure, which is likely a thing in some COTS systems, we are potentially causing a bug that they can't fix.

Reply:
(This is my long-winded way of saying if I didn't recuse myself, I would approve and merge this PR 😄 )

Reply:
Ok, I'll let this one stay as is for now, if it ever comes up that this needs to be fixed, you're on the hook for it 😉

Reply:
I've hardened my position after another thread on the GitLab provider brought up how this `group:` prefix broken htpasswd users in odd ways (they had to hack in the `group:` prefix I think). 😄